narwhal/types/src/primary.rs

```diff
@@ -30,9 +31,61 @@ use std::{
 /// The round number.
 pub type Round = u64;
 
+/// The epoch UNIX timestamp in milliseconds
+pub type TimestampMs = u64;
+
+pub trait Timestamp {
+    fn elapsed(&self) -> Duration;
+}
+
+impl Timestamp for TimestampMs {
+    // Returns the time elapsed between the timestamp
+    // and "now". The result is a Duration.
+    fn elapsed(&self) -> Duration {
+        let diff = now().saturating_sub(*self);
+        Duration::from_millis(diff)
+    }
+}
+// Returns the current time expressed as UNIX
+// timestamp in milliseconds
+fn now() -> TimestampMs {
+    match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
+        Ok(n) => n.as_millis() as TimestampMs,
+        Err(_) => panic!("SystemTime before UNIX EPOCH!"),
+    }
+}
+
+// Additional metadata information for an entity. Those data
+// should not be treated as trustworthy data and should be used
+// for NON CRITICAL purposes only.
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Arbitrary, MallocSizeOf)]
+pub struct Metadata {
+    // timestamp of when the entity created. This is generated
+    // by the node which creates the entity.
+    pub timestamp: TimestampMs,
```

Comment:
Maybe call this creation_time_ms?

Reply:
To be honest I was going a bit back and forth on the naming of this - I was thinking including the `creation` as well. But that metadata refer to immutable data structs which we never modify. So in the end thought of the `creation` might be a bit redundant - but I am open to rename if we think that as important

Reply:
I was about to make the same suggestion and then saw this comment was already here :) FWIW I also think the field name should include something that indicates "timestamp of what?"

Reply:
Thanks both for your comments on this. Ok I'll rename this and put in the mix the "created" name since it seems it will make things clearer